lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,254 @@
+    if (EndTDeallocatesCoarray()) {
+      messages_.Say(charBlock_,
+          "implicit deallocation of coarray not allowed"
+          " in DO CONCURRENT"_err_en_US);
+    }
+  }
+  // C1141: cannot call ieee_get_flag, ieee_[gs]et_halting_mode
+  void Post(const parser::ProcedureDesignator &procedureDesignator) {
+    if (auto *name{std::get_if<parser::Name>(&procedureDesignator.u)}) {
+      // C1137: call move_alloc with coarray arguments
+      if (name->ToString() == "move_alloc"s) {
+        if (anyObjectIsCoarray()) {
+          messages_.Say(charBlock_,
+              "call to MOVE_ALLOC intrinsic in DO CONCURRENT with coarray"
+              " argument(s) not allowed"_err_en_US);
+        }
+      }
+      // C1139: call to impure procedure
+      if (name->symbol && !isPure(name->symbol->attrs())) {
+        messages_.Say(charBlock_,
+            "call to impure subroutine in DO CONCURRENT not allowed"_err_en_US);
+      }
+      if (name->symbol && fromScope(*name->symbol, "ieee_exceptions"s)) {
+        if (name->ToString() == "ieee_get_flag"s) {
+          messages_.Say(charBlock_,
+              "IEEE_GET_FLAG not allowed in DO CONCURRENT"_err_en_US);
+        } else if (name->ToString() == "ieee_set_halting_mode"s) {
+          messages_.Say(charBlock_,
+              "IEEE_SET_HALTING_MODE not allowed in DO CONCURRENT"_err_en_US);
+        } else if (name->ToString() == "ieee_get_halting_mode"s) {
+          messages_.Say(charBlock_,
+              "IEEE_GET_HALTING_MODE not allowed in DO CONCURRENT"_err_en_US);
+        }
+      }
+    } else {
+      // C1139: this a procedure component
+      auto &component{std::get<parser::ProcComponentRef>(procedureDesignator.u)
+                          .v.thing.component};
+      if (component.symbol && !isPure(component.symbol->attrs())) {
+        messages_.Say(charBlock_,
+            "call to impure subroutine in DO CONCURRENT not allowed"_err_en_US);
+      }
+    }
+  }
+
+  // 11.1.7.5
+  void Post(const parser::IoControlSpec &ioControlSpec) {
+    if (auto *charExpr{
+            std::get_if<parser::IoControlSpec::CharExpr>(&ioControlSpec.u)}) {
+      if (std::get<parser::IoControlSpec::CharExpr::Kind>(charExpr->t) ==
+          parser::IoControlSpec::CharExpr::Kind::Advance) {
+        messages_.Say(charBlock_,
+            "ADVANCE specifier not allowed in DO CONCURRENT"_err_en_US);
+      }
+    }
+  }
+
+private:
+  bool anyObjectIsCoarray() { return false; }  // placeholder
+  bool anyObjectIsPolymorphic() { return false; }  // placeholder
+  bool EndTDeallocatesCoarray() { return false; }  // placeholder
+  bool isPure(Attrs &attrs) {
+    return attrs.test(Attr::PURE) ||
+        (attrs.test(Attr::ELEMENTAL) && !attrs.test(Attr::IMPURE));
+  }
+  bool fromScope(const Symbol &symbol, const std::string &moduleName) {
+    if (symbol.scope() && symbol.scope()->IsModule()) {
+      if (symbol.scope()->symbol()->name().ToString() == moduleName) {
```

Comment:
`symbol.scope()` is the scope introduced by a symbol. `symbol.owner()` is the scope that it is in. Also, you need to call `GetUltimate()` to go from the local symbol to the one it is use-associated with. So the body of this function should be something like this:
```
    auto &owner{symbol.GetUltimate().owner()};
    return owner.IsModule() && owner.name().ToString() == moduleName;
```

Reply:
Thanks. Ultimate meant nothing to me.